### ISSUE TITLE
Conditional labeling error in geom_text

### ISSUE THREAD

**Issue by kimchpekr** [tidyverse/ggplot2#2105]
i tried conditional labeling using geom_text like belows

1. this works well
`> ggplot(mpg, aes(class)) + geom_bar() + geom_text(aes(label = ifelse(..count.. < 20, ..count.., "")), stat="count", vjust=-1)`

the result like below.
![img](https://cloud.githubusercontent.com/assets/22344554/25000721/505f4dd6-207d-11e7-8d6f-7a43e8b85c1d.png)

2. this not works and issues error
`> ggplot(mpg, aes(class)) + geom_bar() + geom_text(aes(label = ifelse(class == "compact", ..count.., "")), stat="count", vjust=-1)`

> Error in class == "compact" : 
  comparison (1) is possible only for atomic and list types

i can't understand it. is this bug? or my mistake?
thanks for any helps !

**Comment by has2k1** [comment]
It is your mistake. What you are trying to do is not possible. If any part of the aesthetic is computed (has ..var.. in it), you cannot refer to columns in the original dataframe. That information does not exist.

Here is a hack for you that should work

```R
ggplot(mpg, aes(class)) + geom_bar() + geom_text(aes(label = ifelse(..x.. == "2", ..count.., "")), stat="count", vjust=-1)
```

**Comment by kimchpekr** [comment]
thanks for your comment and hack !
but, i'm confusing still.

you say "That information does not exist".
what this mean?

i think like belows

1) geom_text understans "class" variable so can use this in aes()
`> ggplot(mpg, aes(class)) + geom_bar() + geom_text(aes(label = ifelse(class == "compact", "good", "")), stat="count", vjust=-1) # work`
![img1](https://cloud.githubusercontent.com/assets/22344554/25029618/07d061cc-20fa-11e7-9954-69a56cb999b6.png)

2) stat_count computes count variable so we can use ..count..
`> ggplot(mpg, aes(class)) + geom_bar() + geom_text(aes(label = ifelse(..count.. < 20, ..count.., "")), stat="count", vjust=-1) # work`
![img2](https://cloud.githubusercontent.com/assets/22344554/25029626/1128f1e4-20fa-11e7-972e-633f5f5a0cca.png)


3) as a result, "class", "..count.." variables can be usable. but not.
`> ggplot(mpg, aes(class)) + geom_bar() + geom_text(aes(label = ifelse(class == "compact", ..count.., "")), stat="count", vjust=-1) # not work`

to sum up...
`aes(label = ifelse(class == "compact", "good", "")) # work`
`aes(label = ifelse(..count.. < 20, ..count.., "")) # work`
`aes(label = ifelse(class == "compact", ..count.., "")) # not work`

i can't understand above results. what information does not exist?
i want to understand the reason about this confusing.

thanks for your comments too !

**Comment by has2k1** [comment]
```R
aes(label = ifelse(class == "compact", "good", ""))     # 1. work 
aes(label = ifelse(..count.. < 20, ..count.., ""))      # 2. work
aes(label = ifelse(class == "compact", ..count.., ""))  # 3. not work
```
* `1` - Evaluated **before** computing the statistics
* `2` and `3` - Evaluated **after** computing the statistics

Any mapping with a computed (`..var..`) variable is evaluated after the statistics are computed. At this point ggplot2 has processed the data in different ways and the original columns do not exist. Those that matter have been "standardised" and some information moved to the scales. 

Maybe pseudo-code can help;

```
- If does not have computed variables, evaluate aesthetic
- process data
- ...
- process data (e.g. compute statistics)
- If has computed variables, evaluate aesthetic
```

Think of it this way

1. `geom_text(aes(label = ifelse(class == "compact", "good", "")` - Evaluated when `class` variable exists in the dataframe
2. `geom_text(aes(label = ifelse(..count.. < 20, ..count.., "")` - Evaluated when `..count..` *computed* variable exists in the dataframe. 
3. `geom_text(aes(label = ifelse(class == "compact", ..count.., ""))` - Evaluated when `class` variable does not exist (it has been processed into something else), but the `..count..` *computed* variable exists in the dataframe.

**Comment by kimchpekr** [comment]
i completely understand !
thank you for your helps...
